### PR TITLE
[OV20] ov::Function - get/set batch size

### PR DESCRIPTION
**Pull request by nosovmik**
### Details:
 - Partially reused code from PR #8560 
 - ov::util::get_batch_size - gets batch size from function using layouts of its' parameters (N). In case of error, throws exception with user-friendly explanation of possible reasons
 - ov::util::set_batch_size - sets batch size to function using layouts of its' parameters (N). In case of error, throws exception with user-friendly explanation of possible reasons. Possible errors are same as for 'get_batch_size' + error during validation
 - Test code coverage: 100%

### Tickets:
 - 67997

### CODE REVIEW

File: src/core/include/openvino/core/function.hpp [hunk 1]
```diff
@@ -353,4 +353,43 @@ class OPENVINO_API AttributeAdapter<std::shared_ptr<ov::Function>>
...
+
+/// \brief Helper method to set batch size to a Function
+///
+/// \details Checks layout of each parameter in a Function and sets value for N (B) dimension. Then performs validation
+/// and type propagation
+///
+/// \throws ::ov::AssertFailure with details in case of error. Possible errors are:
+/// * There is no parameter with N dimension in layout. Function shall have at least one parameter with layout with 'N'
+/// dimension. Recommended fix is to use `Parameter::set_layout` API, e.g.
+/// `function->get_parameters[some_index]->set_layout("NCHW");`
```
Comment (ilya-lavrenov) [inline src/core/include/openvino/core/function.hpp:382]: `get_parameters()[]`

File: src/core/src/function.cpp [hunk 2]
```diff
@@ -913,3 +913,112 @@ ov::Output<ov::Node> ov::Function::add_output(const ov::Output<ov::Node>& port)
...
+    }
+    return batch_size;
+}
+
+void ov::util::set_batch_size(const std::shared_ptr<ov::Function>& f, ov::Dimension batch_size) {
+    get_batch_size(f);  // Ensure that function's batch size is valid and can be changed
+    // Now batch size can be set for all needed parameters
+    for (const auto& param : f->get_parameters()) {
+        const auto& layout = param->get_layout();
+        if (layout.empty() || !ov::layout::has_batch(layout))
```
Comment (ilya-lavrenov) [inline src/core/src/function.cpp:1000]: why do we need additionally check for `empty`? I suppose `!ov::layout::has_batch(layout)` should be enough?

File: src/core/src/function.cpp [hunk 3]
```diff
@@ -913,3 +913,112 @@ ov::Output<ov::Node> ov::Function::add_output(const ov::Output<ov::Node>& port)
...
+        stream << "Validation fails after set_batch_size. Possible reasons are:" << std::endl;
+        stream << "    1) Ensure that all inputs have valid layout set with batch dimension" << std::endl;
+        stream << "    2) Check model's documentation if batch size can be set to it at all" << std::endl;
+        stream << "Available inputs after set_batch_size:" << std::endl;
+        for (size_t i = 0; i < f->get_parameters().size(); ++i) {
+            bs_util::dump_parameter(stream, f, i);
+        }
+        stream << "---" << std::endl;
+        stream << "Original error message is: " << e.what();
+        OPENVINO_ASSERT(false, stream.str());
```
Comment (ilya-lavrenov) [inline src/core/src/function.cpp:1022]: I suppose we need to restore original shapes like we do in reshape? See https://github.com/openvinotoolkit/openvino/blob/master/src/core/src/function.cpp#L865-L867

BTW, why this method is not implemented via reshape? It calls smart reshape as well

File: src/core/src/function.cpp (hunk 2 above)
Comment (nosovmik) [reply src/core/src/function.cpp]: Fixed, thanks, missed this from original PoC's code

File: src/core/src/function.cpp (hunk 3 above)
Comment (nosovmik) [reply src/core/src/function.cpp:1032]: Good suggestion, added reusing of 'reshape'. Also enhanced error message to show original shape and intended shape after setting batch size

**Review by ilya-lavrenov — APPROVED** [review]

File: src/core/src/function.cpp [hunk 4]
```diff
@@ -913,3 +913,122 @@ ov::Output<ov::Node> ov::Function::add_output(const ov::Output<ov::Node>& port)
...
+        stream << "---" << std::endl;
+        stream << "Please use 'set_layout' API to set layout with batch dimension, e.g. "
+                  "`Function->get_parameters()[index]->set_layout(\"NCHW\");`";
+
+        OPENVINO_ASSERT(false, stream.str());
+    }
+    return batch_size;
+}
+
+void ov::util::set_batch_size(const std::shared_ptr<ov::Function>& f, ov::Dimension batch_size) {
```
Comment (ilya-lavrenov) [inline src/core/src/function.cpp:995]: BTW, can we reuse it in existing samples like https://github.com/openvinotoolkit/openvino/blob/master/samples/cpp/ngraph_function_creation_sample/main.cpp#L325

Maybe in a separate PR.

File: src/core/include/openvino/core/function.hpp [hunk 5]
```diff
@@ -353,4 +353,43 @@ class OPENVINO_API AttributeAdapter<std::shared_ptr<ov::Function>>
     OPENVINO_RTTI("AttributeAdapter<std::shared_ptr<Function>");
     BWDCMP_RTTI_DECLARATION;
 };
+
+namespace util {
```
Comment (ilya-lavrenov) [inline src/core/include/openvino/core/function.hpp:357]: why is it `util`?

**Review by ilyachur — APPROVED** [review]

File: src/core/include/openvino/core/function.hpp (hunk 5 above)
Comment (nosovmik) [reply src/core/include/openvino/core/function.hpp:357]: This was from original PR #8560 . @ilyachur can you please comment?

**Comment by jane-intel** [comment]
@nosovmik @ilya-lavrenov @ilyachur do we really need to have word `size` in the methods name?

This already was a bit extra for the API1
But now we return `ov::Dimension` from the `get_batch_size` and it could be dynamic, so the word `size` becomes confusing

File: src/core/include/openvino/core/function.hpp (hunk 5 above)
Comment (nosovmik) [reply src/core/include/openvino/core/function.hpp]: changed to 'pass'

File: src/core/src/function.cpp (hunk 4 above)
Comment (nosovmik) [reply src/core/src/function.cpp]: added to this pr

**Review by ilya-lavrenov — APPROVED** [review]

**Review by jane-intel — APPROVED** [review]

**Pull request merged**